util/src/test/java/io/camunda/zeebe/util/startup/StartupProcessTest.java

```diff
@@ -0,0 +1,382 @@
+import java.util.List;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CountDownLatch;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+@SuppressWarnings({"rawtypes", "unchecked"})
+class StartupProcessTest {
+  private static final Object STARTUP_CONTEXT =
+      new Object() {
+        @Override
+        public String toString() {
+          return "startupContext";
+        }
+      };
+
+  private static final Object SHUTDOWN_CONTEXT =
+      new Object() {
+        @Override
+        public String toString() {
+          return "shutdownContext";
+        }
+      };
+
+  private static final ConcurrencyControl TEST_CONCURRENCY_CONTROL = new TestConcurrencyControl();
+
+  static final class InvocationCountingStartupStep implements StartupStep<Object> {
+
+    private int startupInvocationCounter = 0;
+    private int shutdownInvocationCounter = 0;
+
+    int getStartupInvocationCounter() {
+      return startupInvocationCounter;
+    }
+
+    int getShutdownInvocationCounter() {
+      return shutdownInvocationCounter;
+    }
+
+    @Override
+    public String getName() {
+      return "InvocationCountingStartupStep";
+    }
+
+    @Override
+    public ActorFuture<Object> startup(final Object o) {
+      startupInvocationCounter++;
+      return completedFuture(o);
+    }
+
+    @Override
+    public ActorFuture<Object> shutdown(final Object o) {
+      shutdownInvocationCounter++;
+      return completedFuture(o);
+    }
+  }
+
+  static final class WaitingStartupStep implements StartupStep<Object> {
+
+    private final CountDownLatch startupCountdownLatch;
+    private final boolean completeWithException;
+
+    WaitingStartupStep(
+        final CountDownLatch startupCountdownLatch, final boolean completeWithException) {
+      this.startupCountdownLatch = startupCountdownLatch;
+      this.completeWithException = completeWithException;
+    }
+
+    @Override
+    public String getName() {
+      return "WaitingStartupStep";
+    }
+
+    @Override
+    public ActorFuture<Object> startup(final Object o) {
+      final var startupFuture = new TestActorFuture<>();
+      final var startupThread =
+          new Thread(
+              () -> {
+                try {
+                  startupCountdownLatch.await();
+                } catch (final InterruptedException e) {
+                  e.printStackTrace();
+                } finally {
+                  if (!completeWithException) {
+                    startupFuture.complete(o);
+                  } else {
+                    startupFuture.completeExceptionally(new Throwable("completed exceptionally"));
+                  }
+                }
+              });
+      startupThread.start();
+      return startupFuture;
+    }
+
+    @Override
+    public ActorFuture<Object> shutdown(final Object o) {
+      return completedFuture(o);
+    }
+  }
+
+  @Nested
+  class MainUseCase {
```

Comment:
:question: I think this is uncommon in our code base, using `@Nested` and inner classes to separate the use case kinds. Typically we just have different test suites. Can you explain a bit what you like about this? I'm still figuring out how to best use junit 5 as well, so this is useful :slightly_smiling_face:

Reply:
I am also unsure myself and experimenting. 

Basically I do a bit of TDD and then I often start with very uninteresting test cases, like "What if I pass null in the constructor?". Since these are the first test cases I create they are usually at the top of the test class, but also boring to read.

So that is why I like to lump them all up into a nested test case. And then I have another nested class to indicate: Here is the interesting stuff!